crates/bls12381/src/curves/g1.rs

```diff
@@ -469,7 +568,112 @@ mod tests {
         }
         assert!(cs.is_satisfied());
         expect_eq(cs.num_inputs(), &expect!["1"]);
-        expect_eq(cs.scalar_aux().len(), &expect!["8913"]);
-        expect_eq(cs.num_constraints(), &expect!["8919"]);
+        expect_eq(cs.scalar_aux().len(), &expect!["4769"]);
+        expect_eq(cs.num_constraints(), &expect!["4741"]);
+    }
+
+    #[test]
+    fn test_random_mul_by_seed_square() {
+        let mut rng = rand::thread_rng();
+        let a = G1Projective::random(&mut rng);
+        let x0 = bls12_381::Scalar::from(15132376222941642752);
+        let c = a * (x0 * x0);
+        let a = G1Affine::from(a);
+        let c = G1Affine::from(c);
+
+        let mut cs = TestConstraintSystem::<Fp>::new();
+        let a_alloc = G1Point::alloc_element(&mut cs.namespace(|| "alloc a"), &a).unwrap();
+        let c_alloc = G1Point::alloc_element(&mut cs.namespace(|| "alloc c"), &c).unwrap();
+        let res_alloc = a_alloc
+            .scalar_mul_by_seed_square(&mut cs.namespace(|| "a.mul_by_seed_square()"))
+            .unwrap();
+        G1Point::assert_is_equal(
+            &mut cs.namespace(|| "a.mul_by_seed_square() = c"),
+            &res_alloc,
+            &c_alloc,
+        )
+        .unwrap();
+        if !cs.is_satisfied() {
+            eprintln!("{:?}", cs.which_is_unsatisfied())
+        }
+        assert!(cs.is_satisfied());
+        expect_eq(cs.num_inputs(), &expect!["1"]);
+        expect_eq(cs.scalar_aux().len(), &expect!["489860"]);
+        expect_eq(cs.num_constraints(), &expect!["491128"]);
+    }
+
+    #[test]
+    fn test_random_subgroup_check() {
+        let mut rng = rand::thread_rng();
+        let n = Scalar::random(&mut rng);
+        let a = G1Affine::from(G1Projective::generator() * n);
```

Comment:
This is not meant as criticism, but your testing life would I feel be a lot easier with `proptest`, specifically with its closure-style invocation:
https://docs.rs/proptest/latest/proptest/macro.proptest.html#closure-style-invocation

I'm in favor of adding instances of Arbitrary for core types wherever useful, and for foreign types, you can look at how Arecibo does it here:
https://github.com/lurk-lab/arecibo/blob/74792f9324e317020a11d265d7379aca7ca99f66/src/r1cs/util.rs

Reply:
Yeah, I think you're very right. I haven't gotten around to integrating proptest yet but I think implementing Arbitrary for the EC types would've made my life easier, I'll probably do it before the `hash_to_g2` PR is finished